# Homework 5

## Cross-Validation, Confusion Matrices, and Confidence Intervals

<p>
The dataset I chose is the iris dataset
</p>

3 Different models with low, medium, and high complexity

In [1]:
import pandas as pd
from treeviz import tree_print
from sklearn import tree
from sklearn.metrics import accuracy_score
# sklearn provides manipulation of training sets
# here we do train/test split
from sklearn.model_selection import train_test_split

# set up our sklearn data shape for the iris data
df = pd.read_csv("https://raw.githubusercontent.com/IndraniMandal/ds-assets/main/assets/iris.csv")
X  = df.drop(['id','Species'],axis=1)
y = df['Species']

# split the data - 70% training 30% testing
datasets = train_test_split(X, y, train_size=0.7, test_size=0.3, random_state=2)
(X_train, X_test, y_train, y_test) = datasets

Low Complexity Tree:

In [2]:
# set up the tree model object 
model = tree.DecisionTreeClassifier(criterion='entropy', max_depth=1)

# fit the model on the training set of data
model.fit(X_train, y_train)
tree_print(model,X)

# Train results: evaluate the model on the testing set of data
y_train_model = model.predict(X_train)
print("Train Accuracy: {:3.2f}".format(accuracy_score(y_train, y_train_model)))

# Test results: evaluate the model on the testing set of data
y_test_model = model.predict(X_test)
print("Test Accuracy: {:3.2f}".format(accuracy_score(y_test, y_test_model)))

if Petal.Length =< 2.350000023841858: 
  |then setosa
  |else virginica
<->
Tree Depth:  1
Train Accuracy: 0.67
Test Accuracy: 0.67


Medium Complexity Tree:

In [3]:
# set up the tree model object
model = tree.DecisionTreeClassifier(criterion='entropy', max_depth=3)

# fit the model on the training set of data
model.fit(X_train, y_train)
tree_print(model,X)

# Train results: evaluate the model on the testing set of data
y_train_model = model.predict(X_train)
print("Train Accuracy: {:3.2f}".format(accuracy_score(y_train, y_train_model)))

# Test results: evaluate the model on the testing set of data
y_test_model = model.predict(X_test)
print("Test Accuracy: {:3.2f}".format(accuracy_score(y_test, y_test_model)))

if Petal.Length =< 2.350000023841858: 
  |then setosa
  |else if Petal.Width =< 1.6500000357627869: 
  |  |then if Petal.Length =< 4.950000047683716: 
  |  |  |then versicolor
  |  |  |else virginica
  |  |else if Petal.Length =< 4.8500001430511475: 
  |  |  |then virginica
  |  |  |else virginica
<------->
Tree Depth:  3
Train Accuracy: 0.98
Test Accuracy: 0.98


High Complexity Tree:

In [4]:
# set up the tree model object
model = tree.DecisionTreeClassifier(criterion='entropy', max_depth=None)

# fit the model on the training set of data
model.fit(X_train, y_train)
tree_print(model,X)

# Train results: evaluate the model on the testing set of data
y_train_model = model.predict(X_train)
print("Train Accuracy: {:3.2f}".format(accuracy_score(y_train, y_train_model)))

# Test results: evaluate the model on the testing set of data
y_test_model = model.predict(X_test)
print("Test Accuracy: {:3.2f}".format(accuracy_score(y_test, y_test_model)))

if Petal.Width =< 0.800000011920929: 
  |then setosa
  |else if Petal.Width =< 1.6500000357627869: 
  |  |then if Petal.Length =< 4.950000047683716: 
  |  |  |then versicolor
  |  |  |else if Petal.Width =< 1.550000011920929: 
  |  |  |  |then virginica
  |  |  |  |else versicolor
  |  |else if Petal.Length =< 4.8500001430511475: 
  |  |  |then if Sepal.Width =< 3.100000023841858: 
  |  |  |  |then virginica
  |  |  |  |else versicolor
  |  |  |else virginica
<---------->
Tree Depth:  4
Train Accuracy: 1.00
Test Accuracy: 0.96


Looking at the three models, we see that they each have the following test accuracies:
<ul>
<li>low-complexity - 0.67
<li>med-complexity - 0.98
<li>high-complexity - 0.96
</ul><br>

Looking at the training accuracyies, we see the following values:
<ul>
<li>Low-Complexity - 0.67
<li>Med-Complexity - 0.98
<li>High-Complexity - 1.0
</ul><br>
From this we can see that the medium complexity model performed the best
The low-complexity model was underfitting, the medium and high complexity models were neither over or underfitting.

Split the dataset

In [5]:
import pandas as pd
import numpy as np
np.set_printoptions(formatter={'float_kind':"{:3.2f}".format})
from sklearn import tree
# import kFold and cross validation score
from sklearn.model_selection import KFold, cross_val_score


# Get the data
df = pd.read_csv("https://raw.githubusercontent.com/IndraniMandal/ds-assets/main/assets/iris.csv")
features = df.drop(['id', 'Species'], axis=1)
target = df['Species']

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(features, target, test_size=0.2, shuffle=True, random_state=42)

Model building with cross validation: low-complexity

In [7]:
# Do the 5-fold cross validation and shuffle the data around
cv = KFold(n_splits=5, shuffle = True)

# set up the model
dtree = tree.DecisionTreeClassifier(criterion='entropy', max_depth=1)

# Do the 5-fold cross validation
scores = cross_val_score(dtree, X_train, y_train, cv=cv)
print("Fold Accuracies: {}".format(scores))
print("Accuracy: {:3.2f}".format(scores.mean()))

Fold Accuracies: [0.50 0.67 0.58 0.71 0.62]
Accuracy: 0.62


Medium-Complexity

In [8]:
# Do the 5-fold cross validation and shuffle the data around
cv = KFold(n_splits=5, shuffle = True)

# set up the model
dtree = tree.DecisionTreeClassifier(criterion='entropy', max_depth=3)

# Do the 5-fold cross validation
scores = cross_val_score(dtree, X_train, y_train, cv=cv)
print("Fold Accuracies: {}".format(scores))
print("Accuracy: {:3.2f}".format(scores.mean()))

Fold Accuracies: [0.92 0.96 0.96 0.88 1.00]
Accuracy: 0.94


High Complexity:

In [9]:
# Do the 5-fold cross validation and shuffle the data around
cv = KFold(n_splits=5, shuffle = True)

# set up the model
dtree = tree.DecisionTreeClassifier(criterion='entropy', max_depth=5)

# Do the 5-fold cross validation
scores = cross_val_score(dtree, X_train, y_train, cv=cv)
print("Fold Accuracies: {}".format(scores))
print("Accuracy: {:3.2f}".format(scores.mean()))

Fold Accuracies: [0.96 0.92 0.96 0.83 0.88]
Accuracy: 0.91


Grid search with cross-validation

In [11]:
import pandas as pd
from sklearn import tree
from sklearn.metrics import accuracy_score
from sklearn.model_selection import GridSearchCV

# get data
df = pd.read_csv('https://raw.githubusercontent.com/IndraniMandal/ds-assets/main/assets/iris.csv')
X  = df.drop(['id','Species'],axis=1)
y = df['Species']

# setting up grid search
model = model = tree.DecisionTreeClassifier(random_state=1)
param_grid = {
    'max_depth': list(range(1,11)), # search 1..10
    'criterion': ['entropy', 'gini']
    }
grid = GridSearchCV(model, param_grid, cv=5)

# performing grid search 
grid.fit(X,y)

# print out best parameters
print("Best parameters: {}".format(grid.best_params_))

# print out the best model
print("Best tree:")
tree_print(grid.best_estimator_,X)

# compute the accuracy of optimal classifier      
predict_y = grid.best_estimator_.predict(X)
acc = accuracy_score(y, predict_y)

# print accuracy          
print("Accuracy of optimal classifier: {:3.2f}".format(acc))

Best parameters: {'criterion': 'gini', 'max_depth': 4}
Best tree:
if Petal.Width =< 0.800000011920929: 
  |then setosa
  |else if Petal.Width =< 1.75: 
  |  |then if Petal.Length =< 4.950000047683716: 
  |  |  |then if Petal.Width =< 1.6500000357627869: 
  |  |  |  |then versicolor
  |  |  |  |else virginica
  |  |  |else if Petal.Width =< 1.550000011920929: 
  |  |  |  |then virginica
  |  |  |  |else versicolor
  |  |else if Petal.Length =< 4.8500001430511475: 
  |  |  |then if Sepal.Length =< 5.950000047683716: 
  |  |  |  |then versicolor
  |  |  |  |else virginica
  |  |  |else virginica
<---------->
Tree Depth:  4
Accuracy of optimal classifier: 0.99


The best model parameters from grid search are:
<ul>
<li> criterion : gini
<li> max_depth : 4
</ul>

Confidence Interval Calculations:

In [17]:
def classification_confint(acc, n):
    '''
    Compute the 95% confidence interval for a classification problem.
      acc -- classification accuracy
      n   -- number of observations used to compute the accuracy
    Returns a tuple (lb,ub)
    '''
    import math
    interval = 1.96*math.sqrt(acc*(1-acc)/n)
    lb = max(0, acc - interval)
    ub = min(1.0, acc + interval)
    return (lb,ub)

def regression_confint(rs_score, n, k):
    '''
    Compute the 95% confidence interval for a regression problem.
      rs_score -- R^2 score
      n        -- number of observations used to compute the R^2 score
      k        -- number of independent variables in dataset
    Returns a tuple (lb,ub)
    Reference:
    https://books.google.com/books?id=gkalyqTMXNEC&pg=PA88#v=onepage&q&f=false
    '''
    import math
    interval = 2*math.sqrt((4*rs_score*(1-rs_score)**2*(n-k-1)**2)/((n**2 - 1)*(n+3)))
    lb = max(0, rs_score - interval)
    ub = min(1.0, rs_score + interval)
    return (lb,ub)

# cross-validation Iris
import pandas as pd
import numpy as np
np.set_printoptions(formatter={'float_kind':"{:3.2f}".format})
from sklearn import tree
# grab cross validation code
from sklearn.model_selection import cross_val_score, KFold,train_test_split 

# get data
url = 'https://raw.githubusercontent.com/IndraniMandal/ds-assets/main/assets/iris.csv' # the URL
df = pd.read_csv(url)
features  = df.drop(['id','Species'],axis=1)
target = df['Species']

#splitting the dataset into train and test sets
X_train, X_test, y_train, y_test = train_test_split(features, target, train_size=0.8, test_size=0.2, random_state=2)

# do the 5-fold cross validation and shuffle the data
cv = KFold(n_splits=5,  shuffle = True)

# set up the model
dtree = tree.DecisionTreeClassifier(criterion='entropy', max_depth=4, random_state=1)

dtree.fit(X_train, y_train)

pred_test = dtree.predict(X_test)

acc = accuracy_score(y_test, pred_test)
observations = X_test.shape[0]

# do the 5-fold cross validation
scores = cross_val_score(dtree, X_train, y_train,cv=cv)
print("Fold Accuracies: {}".format(scores))

lb,ub = classification_confint(acc, observations)
print("Accuracy: {:3.2f} ({:3.2f},{:3.2f})".format(acc,lb,ub))

Fold Accuracies: [1.00 0.92 0.88 0.88 1.00]
Accuracy: 0.93 (0.84,1.00)


The confidence interval is 0.84 to 1.00

The medium complexity model and the grid-search optimal model overlap highly with their cross-validation scores. I would choose the medium-complexity model as it has a higher accuracy at 0.94 whereas the grid-search model is an accuracy of 0.93